fuzzers/main.py

```diff
@@ -0,0 +1,842 @@
+```
+cd {quoted_crate}
+RUSTC_BOOTSTRAP=1 cargo fuzz tmin {quoted_runner} {quoted_artifact}
+```
+
+Please edit the topic name to add more meaningful information once investigated. \
+Keeping the artifact hash in it can help if the same artifact gets detected as \
+crashing another branch.
+'''
+        })
+        focus_log_lines = '\n'.join(log_lines[:5]) + '\n[...]\n'
+        for line in log_lines[::-1]:
+            if line.startswith('```'):
+                # Censor the end of a spoiler block, not great but this is for human consumption
+                # anyway
+                line = ' ' + line
+            if len(focus_log_lines) + len(line) > 9000:
+                # Zulip limit is 10k, let's keep some safety buffer here
+                break
+            focus_log_lines = line + focus_log_lines
+        client.send_message({
+            'type': 'stream',
+            'to': 'pagoda/fuzzer/private',
+            'topic': f'{branch}: artifact {artifact}',
+            'content': f'```spoiler First and last few log lines\n{focus_log_lines}\n```',
+        })
+
+    def signal(self, sig: int) -> None:
+        """
+        Signal the fuzzer process (with all its process group as a fuzzer can spawn sub-fuzzers)
+        """
+
+        print(f'Sending signal {sig} to fuzzer {self.proc.pid}', file=sys.stderr)
+        os.killpg(os.getpgid(self.proc.pid), sig)
+
+
+# Actually this should also require a ['weight']: int bound but it seems like a mess to encode
+T = typing.TypeVar('T')  # pylint: disable=invalid-name
+
+
+def random_weighted(array: typing.List[T], num: int, name: str) -> typing.List[T]:
+    print(f'Picking {num} random {name} among {array}', file=sys.stderr)
+    untyped_array = typing.cast(typing.Any, array)
+    res = random.choices(array, [x['weight'] for x in untyped_array], k=num)
+    print(f' -> picked {res}', file=sys.stderr)
+    return res
+
+
+def pause_exit_spot(pause_evt: threading.Event, resume_evt: threading.Event,
+                    exit_evt: threading.Event) -> bool:
+    """
+    Poll the events, returning True if `exit_evt` was set and pausing as requested by `pause_evt`
+    and `resume_evt`
+    """
+
+    resume_evt.clear()
+    if pause_evt.is_set():
+        while not resume_evt.wait(timeout=1):
+            if exit_evt.is_set():
+                return True
+        pause_evt.clear()
+    return exit_evt.is_set()
+
+
+def kill_fuzzers(bucket: gcs.Bucket,
+                 fuzzers: typing.Iterable[FuzzProcess]) -> None:
+    for fuzzer in fuzzers:
+        try:
+            fuzzer.signal(signal.SIGTERM)
+        except ProcessLookupError:
+            print(f'Failed looking up process {fuzzer.proc.pid}',
+                    file=sys.stderr)
+    time.sleep(5)
+    for fuzzer in fuzzers:
+        try:
+            fuzzer.signal(signal.SIGKILL)
+        except ProcessLookupError:
+            print(f'Failed looking up process {fuzzer.proc.pid}',
+                    file=sys.stderr)
```

Comment:
At this point this pretty much expected though, no? So this should just be `pass` IMO.